apps/managedidentity/managedidentity_test.go

```diff
@@ -4,43 +4,399 @@ package managedidentity
+	}
+	jsonResponse, _ := json.Marshal(response)
+	return string(jsonResponse)
+}
 
-	if err == nil {
-		t.Errorf("TestManagedIdentity: unexpected nil error from TestManagedIdentity")
+func getMsiErrorResponseNoRetry() string {
+	response := ErrorResponse{
+		StatusCode:    123,
+		Message:       "Not one of the retryable error responses",
+		CorrelationID: "7d0c9763-ff1d-4842-a3f3-6d49e64f4513",
 	}
+	jsonResponse, _ := json.Marshal(response)
+	return string(jsonResponse)
 }
 
-func TestManagedIdentityWithClaims(t *testing.T) {
-	client, err := fakeClient(ClientID("123"))
+func computeUri(endpoint string, queryParameters map[string][]string) string {
+	if len(queryParameters) == 0 {
+		return endpoint
+	}
+
+	queryString := url.Values{}
+	for key, values := range queryParameters {
+		for _, value := range values {
+			queryString.Add(key, value)
+		}
+	}
+
+	return endpoint + "?" + queryString.Encode()
+}
+
+func expectedRequest(source Source, resource string, id ID) (*http.Request, error) {
+	return expectedRequestWithId(source, resource, id)
+}
+
+func expectedRequestWithId(_ Source, resource string, id ID) (*http.Request, error) {
+	var endpoint string
+	headers := http.Header{}
+	queryParameters := make(map[string][]string)
+
+	//check with source when added different sources.
+	endpoint = imdsEndpoint
+	queryParameters["api-version"] = []string{"2018-02-01"}
+	queryParameters["resource"] = []string{resource}
+	headers.Add("Metadata", "true")
+
+	switch id.(type) {
+	case ClientID:
+		queryParameters[miQuerryParameterClientId] = []string{id.value()}
+	case ResourceID:
+		queryParameters[miQuerryParameterResourceId] = []string{id.value()}
+	case ObjectID:
+		queryParameters[miQuerryParameterObjectId] = []string{id.value()}
+	case systemAssignedValue:
+		// not adding anything
+	default:
+		return nil, fmt.Errorf("Type not supported")
+	}
 
+	uri, err := url.Parse(computeUri(endpoint, queryParameters))
 	if err != nil {
-		t.Fatal(err)
+		return nil, err
 	}
 
-	_, err = client.AcquireToken(context.Background(), "scope", WithClaims("claim"))
+	req := &http.Request{
+		Method: "GET",
+		URL:    uri,
+		Header: headers,
+	}
+
+	return req, nil
+}
+
+func ExpectedResponse(statusCode int, response string) http.Response {
+	return http.Response{
+		StatusCode: statusCode,
+		Body:       io.NopCloser(strings.NewReader(response)),
+	}
+}
+
+type resourceTestData struct {
+	source   Source
+	endpoint string
+	resource string
+}
+
+func createResourceData() []resourceTestData {
+	return []resourceTestData{
+		{source: DefaultToIMDS, endpoint: imdsEndpoint, resource: resource},
+		{source: DefaultToIMDS, endpoint: imdsEndpoint, resource: resourceDefaultSuffix},
+	}
+}
+
+func Test_SystemAssigned_Returns_Token_Failure(t *testing.T) {
+	testCases := []ErrorResponse{
+		{StatusCode: 404, Message: "IMDS service not available", CorrelationID: "121212"},
+		{StatusCode: 501, Message: "Service error 1", CorrelationID: "121212"},
```

Comment:
There is no message 404 or 5xx errors, only with 400. Please add a test for this.

Reply:
I updated the error testing like in Azure sdk. 
This also tests for service errors.